### PR TITLE
flooder: single connection instead of connection-per-tx

### PR DESCRIPTION
**Pull request by fixxxedpoint**

### CODE REVIEW

**Review by pmikolajczyk41 — CHANGES_REQUESTED** [review]
Ideally, I would split this PR into two smaller ones: one for resolving dependencies and one changing semantics

File: common/Cargo.toml [hunk 1]
```diff
@@ -5,12 +5,15 @@ edition = "2021"
 
 # client
 [dependencies.substrate-api-client]
-git = "https://github.com/scs/substrate-api-client"
-rev = "aa4b84f8775f972e5967aca3d96dd163619f65d5"
+git = "https://github.com/Cardinal-Cryptography/substrate-api-client.git"
+rev = "8b7346a11a70f46d05de543a6bed37f3fdf6ac8f"
```
Comment (pmikolajczyk41) [inline common/Cargo.toml:9]: Unfortunately this SHA corresponds to a latest commit at branch called `single_threaded_sender.PR`, which (pardon me) definitely does not sound like a source of a stable version for 3rd-party library :P
The easiest way is to rename the branch or create a tag.

Comment (fixxxedpoint) [reply common/Cargo.toml:9]: How is renaming of a branch going to fix anything :D

Comment (fixxxedpoint) [reply common/Cargo.toml:9]: I will make a tag of from it, since pushing it to main doesn't look like a good idea.

Comment (fixxxedpoint) [reply common/Cargo.toml]: changed to tag

File: common/src/ws_rpc_client.rs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            _ => Err(ApiClientError::UnsupportedXtStatus(exit_on)),
+        };
+
+        // reset the RpcClient handler used by the WebSocket's thread
+        *self
+            .next_handler
+            .lock()
+            .expect("unable to acquire a lock on RpcClient") = None;
+        result
+    }
```
Comment (pmikolajczyk41) [inline common/src/ws_rpc_client.rs:124]: maybe just add a method like `reset_handler()`

File: common/src/ws_rpc_client.rs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+    fn send_rpc_request(
+        &self,
+        jsonreq: String,
+        result_in: ThreadOut<String>,
+        on_message_fn: OnMessageFn,
+    ) -> WsResult<()> {
+        // 1 used by `on_open` of RpcClient + 1 for `close`
+        const MAGIC_SEND_CONST: usize = 2;
+        let (ws_tx, _ws_rx) = mio::channel::sync_channel(MAGIC_SEND_CONST);
```
Comment (pmikolajczyk41) [inline common/src/ws_rpc_client.rs:176]: `channel` is deprecated, but you can use `mio-extras`

File: common/src/ws_rpc_client.rs [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+            on_extrinsic_msg_until_ready, on_get_request_msg, OnMessageFn, RpcClient,
+        },
+    },
+    ApiClientError, ApiResult, FromHexString, RpcClient as RpcClientTrait, XtStatus,
+};
+use ws::{connect, Handler, Message, Result as WsResult, Sender as WsSender};
+
+pub struct WsRpcClient {
+    mux: Mutex<()>,
+    next_handler: Arc<Mutex<Option<RpcClient>>>,
```
Comment (pmikolajczyk41) [inline common/src/ws_rpc_client.rs:27]: would `Option<Arc<Mutex<RpcClient>>>` work?

**Review by pmikolajczyk41 — APPROVED** [review]
I wasn't able to verify in-depth the mechanics, but seems ok

File: common/src/ws_rpc_client.rs (hunk 3 above)
Comment (fixxxedpoint) [reply common/src/ws_rpc_client.rs:176]: I don't think I can - mismatched types. `Secret` constructor ;) https://docs.rs/ws/0.2.1/src/ws/.cargo/registry/src/github.com-1ecc6299db9ec823/ws-0.2.1/src/communication.rs.html#52

File: common/src/ws_rpc_client.rs (hunk 2 above)
Comment (fixxxedpoint) [reply common/src/ws_rpc_client.rs:124]: Good idea.

**Review by mike1729 — APPROVED** [review]
Very nice work! If the upstream merges the changes from this PR then their average code quality will double :sunglasses:

File: common/src/ws_rpc_client.rs (hunk 4 above)
Comment (fixxxedpoint) [reply common/src/ws_rpc_client.rs:27]: I am sharing this `Option` between threads (`Handler for WsHandler`) so I think I need this `Mutex` around it.

File: common/src/ws_rpc_client.rs (hunk 2 above)
Comment (fixxxedpoint) [reply common/src/ws_rpc_client.rs]: Changed.

**Pull request merged**